feat(product-assistant): enhanced trends generation

skoob13 opened the pull request:
## Problem

A straightforward approach for generating insight schemas is naive. This PR introduces a new strategy for generating schemas through two-step generation:
The ReAct agent for plan generation provides a flexible and robust way to discover suitable taxonomies for events and event property filters.
- The schema generation step using function calling with the Trends JSON schema and the previously generated plan.

## Changes

- Bumped dependencies.
- Bootstraped the agent using LangGraph.
- Moved the two-step generation from the notebooks.

## Does this work well for both Cloud and self-hosted?

No, and it's intentional. The agent is under `ee` and has a different license. We may reconsider this later, but it will be under `ee` for some time.

## How did you test this code?

There are very basic tests for conversation reconstruction. I will add additional tests later.

skoob13 committed:
chore: deps

skoob13 committed:
feat: initial graph

skoob13 committed:
feat: partial toolkit

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

skoob13 committed:
feat: tools

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

skoob13 committed:
feat: tool validation

skoob13 committed:
feat: team query runner

skoob13 committed:
feat: tool loop

skoob13 committed:
feat: trend generation step

skoob13 committed:
refactor: stateful nodes

skoob13 committed:
Merge branch 'feat/trends-agent' of github.com:PostHog/posthog into feat/trends-agent

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

skoob13 committed:
feat: handle unformatted response

skoob13 committed:
feat: pydantic schemas for state

skoob13 committed:
fix: state

skoob13 committed:
feat: initial properties

skoob13 committed:
Merge branch 'feat/trends-agent' of github.com:PostHog/posthog into feat/trends-agent

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

skoob13 committed:
feat: support multiple arguments in tools

skoob13 committed:
feat: retrieve event properties

skoob13 committed:
chore: docs

skoob13 committed:
Merge branch 'feat/trends-agent' of github.com:PostHog/posthog into feat/trends-agent

skoob13 committed:
feat: entity validation in entity prop values retrieval

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

skoob13 committed:
fix: steps joining

skoob13 committed:
feat: session retrieval

skoob13 committed:
Merge branch 'feat/trends-agent' of github.com:PostHog/posthog into feat/trends-agent

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

skoob13 committed:
feat: wip conversations

skoob13 committed:
feat: agent conversation

skoob13 committed:
Merge branch 'master' of github.com:PostHog/posthog into feat/trends-agent

skoob13 committed:
feat: reconstruct message history

skoob13 committed:
Merge branch 'feat/trends-agent' of github.com:PostHog/posthog into feat/trends-agent

skoob13 committed:
chore: __init__.py

skoob13 committed:
chore: execution limits

skoob13 committed:
feat: migrating to a new assistant

skoob13 committed:
feat: streaming (wip)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

Twixes commented on ee/hogai/trends/toolkit.py:
```diff
@@ -0,0 +1,498 @@
+    @cached_property
+    def tools(self) -> list[ToolkitTool]:
+        """
+        Our ReAct agent doesn't use function calling. Instead, it uses tools in natural language to decide next steps. The agent expects the following format:
```
_Should_ it be using function calling?

skoob13 replied:
I don't think so, at least not for the current state of models. There is an [experiment](https://github.com/PostHog/ai_notebooks/blob/main/trends_function_calling.ipynb) that uses function calling for taxonomy discovery, but it doesn't perform as well as ReAct.

One thing that we may explore is combining a ReAct agent and function calling to simplify output parsing while keeping similar <s>reasoning</s> predicting capabilities. Models without function calling aren't fine-tuned for a structured output, so accuracy is lower for a JSON/XML/whatever structured format.

Twixes commented on ee/hogai/trends/nodes.py:
```diff
@@ -0,0 +1,381 @@
+        root.text = (
+            "\n" + "\n".join([f'name "{group.group_type}", index {group.group_type_index}' for group in groups]) + "\n"
+        )
+        return ET.tostring(root, encoding="unicode")
```
Why output XML?

skoob13 replied:
We output XML for data markup. I used lists that were also working, but XMLs or defined separators like XML tags seem to perform much better than any other formats. The trade-off, though, is the increased count of tokens.

skoob13 committed:
refactor: query endpoint

skoob13 committed:
Merge branch 'master' of github.com:PostHog/posthog into feat/trends-agent

skoob13 committed:
feat: streaming

skoob13 committed:
fix: streaming of plans

skoob13 committed:
merge

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

skoob13 committed:
Merge branch 'master' of github.com:PostHog/posthog into feat/trends-agent

skoob13 committed:
chore: sync mypy

skoob13 committed:
chore: sync mypy

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

skoob13 committed:
Revert "chore: sync mypy"

This reverts commit 1e15a3a6cd3340119f6888117b5a61282bfee9a6.

skoob13 committed:
chore: mypy

skoob13 committed:
Merge branch 'feat/trends-agent' of github.com:PostHog/posthog into feat/trends-agent

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

skoob13 committed:
chore: another mypy attempt

skoob13 committed:
Merge branch 'feat/trends-agent' of github.com:PostHog/posthog into feat/trends-agent

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

skoob13 committed:
Merge branch 'master' of github.com:PostHog/posthog into feat/trends-agent

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

skoob13 committed:
fix: toolkit bugs for entity props and values

skoob13 committed:
test: event properties and values

skoob13 committed:
Merge branch 'feat/trends-agent' of github.com:PostHog/posthog into feat/trends-agent

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

skoob13 committed:
chore: prompt tweaks

skoob13 committed:
Merge branch 'feat/trends-agent' of github.com:PostHog/posthog into feat/trends-agent

skoob13 committed:
fix: filter thoughts without an output

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

skoob13 committed:
chore: langfuse for debugging

skoob13 committed:
chore: slightly tweak prompt for math

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

skoob13 committed:
feat: tweak formulas and aggregation

skoob13 committed:
Merge branch 'feat/trends-agent' of github.com:PostHog/posthog into feat/trends-agent

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

skoob13 committed:
fix: deps

skoob13 committed:
Merge branch 'feat/trends-agent' of github.com:PostHog/posthog into feat/trends-agent

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

skoob13 committed:
fix: schema

skoob13 committed:
Merge branch 'feat/trends-agent' of github.com:PostHog/posthog into feat/trends-agent

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update query snapshots

skoob13 committed:
Merge branch 'master' of github.com:PostHog/posthog into feat/trends-agent

skoob13 committed:
Merge branch 'feat/trends-agent' of github.com:PostHog/posthog into feat/trends-agent

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update query snapshots

Twixes committed:
Reorder node metods for readability

Just moving methods around so that `run` is always the first method you read, followed by `router`, and then all the internals. No logic changes.

Twixes commented on ee/hogai/trends/nodes.py:
```diff
@@ -0,0 +1,398 @@
+        output_parser = ReActJsonSingleInputOutputParser()
+        merger = merge_message_runs()
+
+        agent = prompt | merger | ChatOpenAI(model="gpt-4o", temperature=0.7, streaming=True) | output_parser
```
This syntax is just 🤯

Twixes commented on ee/hogai/trends/nodes.py:
```diff
@@ -0,0 +1,398 @@
+        )
+
+    def router(self, state: AssistantState):
+        # Exceptional case. TODO: decide how to handle this.
```
When's the TODO for?

Twixes commented on ee/hogai/trends/toolkit.py:
```diff
@@ -0,0 +1,520 @@
+        runner = EventTaxonomyQueryRunner(EventTaxonomyQuery(event=event_name), self._team)
+        response = runner.run(ExecutionMode.RECENT_CACHE_CALCULATE_BLOCKING_IF_STALE)
+
+        # TODO: incorrect response handling. Should it be retried?
```
When's the TODO for?

Twixes commented on ee/hogai/trends/toolkit.py:
```diff
@@ -0,0 +1,520 @@
+            return f"The property {property_name} does not exist in the taxonomy."
+
+        if property_name == "$channel_type":
+            # Not ideal. TODO: replace later.
```
When's later?

Twixes commented on ee/hogai/trends/nodes.py:
```diff
@@ -0,0 +1,398 @@
+
+    @cached_property
+    def _events_prompt(self) -> str:
+        event_description_mapping = {
+            "$identify": "Identifies an anonymous user. This event doesn't show how many users you have but rather how many users used an account."
+        }
```
This kinda seems like a poorer version of `hardcoded_prop_defs` – why not put this description directly in there?

Twixes reviewed (commented):
Just a few questions on the details, the big picture looks sweet

skoob13 commented on ee/hogai/trends/nodes.py:
```diff
@@ -0,0 +1,398 @@
+                conversation.append(
+                    HumanMessagePromptTemplate.from_template(
+                        react_follow_up_prompt,
+                        template_format="mustache",
```
I primarily used `mustache` for formatting when multiple occurrences of the same variable were represented differently. I later removed it, so using `mustache` has no advantage. Do you prefer the f-string?

Twixes replied:
No preference, was just wondering :)

skoob13 commented on frontend/src/queries/schema.ts:
```diff
@@ -2089,3 +2091,24 @@ export interface ActorsPropertyTaxonomyQuery extends DataNode<ActorsPropertyTaxo
+
+export enum AssistantMessageType {
+    Visualization = 'visualization',
+    /**
+     * Do not use. Force the schema generator to generate a StrEnum.
+     */
+    Noop = 'noop',
```
This value forces the scheme generator to produce a StrEnum rather than a Literal type. I haven't found a way to force the schema builder to create the enum when only one key exists.

Twixes replied:
Ahh, that's fascinating

skoob13 committed:
fix: mypy glitch

skoob13 committed:
Merge branch 'master' of github.com:PostHog/posthog into feat/trends-agent

skoob13 committed:
revert: mypy-baseline

skoob13 committed:
fix: mypy errors after dependencies upgrade

skoob13 commented on mypy-baseline.txt:
It should be fixed now. There were a bunch of errors after upgrading the dependencies.

skoob13 committed:
fix: remove bunch of todos

skoob13 committed:
fix: remove cohorts completely

skoob13 committed:
test: test flow

skoob13 committed:
chore: mypy baseline

skoob13 committed:
fix: do not show langfuse warning

skoob13 committed:
fix: unify hardcoded event mapping

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

skoob13 committed:
fix: distinct messages

skoob13 committed:
fix: send messages

skoob13 committed:
feat: straighforward message inferring on frontend

skoob13 committed:
fix: only stream generation step chunks

skoob13 committed:
fix: thread story

skoob13 committed:
Merge branch 'feat/trends-agent' of github.com:PostHog/posthog into feat/trends-agent

skoob13 commented on ee/hogai/assistant.py:
```diff
@@ -0,0 +1,72 @@
+                    yield FrontendAssistantMessage(
+                        type=message.type, content=message.content, payload=message.payload
+                    ).model_dump_json()
+                elif isinstance(message, AIMessageChunk):
```
Refactored how we handle different message types. It greatly improves DX and much more verbose.

skoob13 commented on mypy-baseline.txt:
```diff
@@ -874,3 +860,16 @@ posthog/api/test/batch_exports/test_update.py:0: error: Value of type "BatchExpo
+posthog/api/test/test_capture.py:0: error: Dict entry 0 has incompatible type "str": "float"; expected "str": "int"  [dict-item]
+posthog/api/test/test_capture.py:0: error: Dict entry 0 has incompatible type "str": "float"; expected "str": "int"  [dict-item]
+posthog/api/test/test_capture.py:0: error: Dict entry 0 has incompatible type "str": "float"; expected "str": "int"  [dict-item]
+posthog/api/test/test_capture.py:0: error: Dict entry 0 has incompatible type "str": "float"; expected "str": "int"  [dict-item]
```
By some reason, the baseline file is sorted in a different order on my machine. I fixed all errors appeared after the Pydantic version bump.

skoob13 commented on posthog/hogql/database/database.py:
```diff
@@ -1,42 +1,42 @@
 import dataclasses
```
I had to update this file to resolve the mypy issues.

skoob13 commented on posthog/hogql/database/test/test_view.py:
Same here

skoob13 commented on posthog/hogql_queries/insights/trends/trends_query_runner.py:
Same here

skoob13 commented on frontend/src/queries/schema.ts:
```diff
@@ -2090,3 +2083,28 @@ export interface ActorsPropertyTaxonomyQuery extends DataNode<ActorsPropertyTaxo
+    answer?: ExperimentalAITrendsQuery
+}
+
+export type RootAssistantMessage = VisualizationMessage | AssistantMessage | HumanMessage
```
If I rename this type to AssistantMessageRoot, the Pydantic generator doesn't use the AssistantMessageType enum🙃 It just replaces the values with the type name, so the type field of AssistantMessage becomes `Literal['AssistantMessage']`.

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update query snapshots

skoob13 committed:
fix: disable flaky test

skoob13 committed:
Merge branch 'feat/trends-agent' of github.com:PostHog/posthog into feat/trends-agent

skoob13 committed:
Merge branch 'master' of github.com:PostHog/posthog into feat/trends-agent

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

skoob13 committed:
Merge branch 'master' of github.com:PostHog/posthog into feat/trends-agent

skoob13 committed:
Merge branch 'feat/trends-agent' of github.com:PostHog/posthog into feat/trends-agent

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

Twixes reviewed (approved):
Let's go

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update query snapshots
